src/daft-local-execution/src/writes/unpartitioned_write.rs

```diff
@@ -0,0 +1,235 @@
+use std::sync::Arc;
+
+use common_display::tree::TreeDisplay;
+use common_error::DaftResult;
+use daft_core::prelude::SchemaRef;
+use daft_micropartition::{FileWriter, MicroPartition};
+use daft_table::Table;
+use snafu::ResultExt;
+
+use super::WriterFactory;
+use crate::{
+    buffer::RowBasedBuffer,
+    channel::{create_channel, PipelineChannel, Receiver, Sender},
+    create_task_set,
+    pipeline::PipelineNode,
+    runtime_stats::{CountingReceiver, RuntimeStatsContext},
+    ExecutionRuntimeHandle, JoinSnafu, TaskSet, NUM_CPUS,
+};
+
+pub(crate) struct UnpartitionedWriteNode {
+    name: &'static str,
+    child: Box<dyn PipelineNode>,
+    runtime_stats: Arc<RuntimeStatsContext>,
+    writer_factory: Arc<dyn WriterFactory>,
+    target_in_memory_file_rows: usize,
+    target_in_memory_chunk_rows: usize,
+    file_schema: SchemaRef,
+}
+
+impl UnpartitionedWriteNode {
+    pub(crate) fn new(
+        name: &'static str,
+        child: Box<dyn PipelineNode>,
+        writer_factory: Arc<dyn WriterFactory>,
+        target_in_memory_file_rows: usize,
+        target_in_memory_chunk_rows: usize,
+        file_schema: SchemaRef,
+    ) -> Self {
+        Self {
+            name,
+            child,
+            runtime_stats: RuntimeStatsContext::new(),
+            writer_factory,
+            target_in_memory_file_rows,
+            target_in_memory_chunk_rows,
+            file_schema,
+        }
+    }
+
+    pub(crate) fn boxed(self) -> Box<dyn PipelineNode> {
+        Box::new(self)
+    }
+
+    // Receives data from the dispatcher and writes it.
+    // If the received file idx is different from the current file idx, this means that the current file is full and needs to be closed.
+    // Once input is exhausted, the current writer is closed and all written file paths are returned.
+    async fn run_writer(
+        mut input_receiver: Receiver<(Arc<MicroPartition>, usize)>,
+        writer_factory: Arc<dyn WriterFactory>,
+    ) -> DaftResult<Vec<Table>> {
+        let mut written_file_paths = vec![];
+        let mut current_writer: Option<Box<dyn FileWriter>> = None;
+        let mut current_file_idx = None;
+        while let Some((data, file_idx)) = input_receiver.recv().await {
+            if current_file_idx.is_none() || current_file_idx.unwrap() != file_idx {
```

Comment:
Reviewed the general pattern of how we are approaching file writes and I think we can improve it with a better abstraction.

So we already have the `trait FileWriter` and `trait FileWriterFactory`

but imagine if we can start layering them.

So at the base we can have `ParquetFileWriter` but instead of implementing the row group batching in the executor you can instead implement `RowBatcherWriter` that takes in a `FileWriterFactory` and then gives you a new factory.

Then you can pass that factory into a `TargetFileSizeWriter` that will target the target file size for each file it's writing out. Finally that factory can be passed into a `PartitionedWriter` that will partition by value.

This pattern is pretty common for building adaptors for writers can be seen here in the Iceberg sdk
https://github.com/apache/iceberg/blob/2b55fef7cc2a249d864ac26d85a4923313d96a59/core/src/main/java/org/apache/iceberg/io/PartitionedWriter.java

The cool thing is that these Filewriters be be parametered at runtime so the executor can likely be very simple.

Reply:
I think I get it. So `PartitionedWriter` is itself a `FileWriter`, which holds a `WriterFactory` which can generate `TargetFileSizeWriter` which is also a `FileWriter` and also holds a `WriterFactory` which can generate a `RowBatchWriter`, which is also a `FileWriter` etc etc etc. 

Implemented this in latest commit, as a blocking sink. Currently is not parallel, I will work on this. Maybe in a separate PR.